### DISCUSSION TITLE
Best practice to synth and release construct and assets to S3 buckets

### DISCUSSION THREAD

**Discussion by mavogel** [projen/projen#3994] [General]
Is there any best practice for a workflow (e.g. GH actions) yet which allows to synth the a developed `awscdk-construct` and publish it to an S3 bucket, including its assets?

I saw some approaches with `CliCredentialsStackSynthesizer` and bash script or using `yarn cdk-assets publish` using the package [cdk-assets](https://www.npmjs.com/package/cdk-assets)

Would this be something to be added to projen?

**Comment by mrgrain** [comment]
> Is there any best practice for a workflow (e.g. GH actions) yet which allows to synth the a developed `awscdk-construct` and publish it to an S3 bucket, including its assets?
> 
> I saw some approaches with `CliCredentialsStackSynthesizer` and bash script or using `yarn cdk-assets publish` using the package [cdk-assets](https://www.npmjs.com/package/cdk-assets)

Not really sure I understand what you want to do here exactly (or why). Do you know how this should look without projen? We can then think about a projen implementation together.

**Reply by mavogel** [reply]
ok let me clarify it a bit more: imagine you develop a `awscdk-construct` and you not only want people to use it via cdk but also use the generated cloudformation template, e.g. via direct link in cloudformation. For example this one [here](https://static.us-east-1.prod.workshops.aws/public/cc4aa67e-5b7a-4df1-abf7-c42502899a25/assets/fleet-workshop-team-stack-self.json)

Questions people might have are:
- how do I generate this json / yaml?
- how do I publish it in my ci pipeline?
- what permission does my IAM role need?
- what permissions do the s3 buckets need to have?
- what do I need to consider regarding the cfn parameter types? Will a work as they do in cdk? e.g passing the object of an S3 bucket or better only the arn?

Things for the docs:
- if my construct contains lambdas, which are used in cfn custom resources and their code is not inline but referenced in a s3 bucket then this bucket needs to be in the same region the stack is executed

My proposal would be to have the answers to the questions encoded in projen. I hope I could clarify it a bit more :) Let me know your thoughts

**Reply by mrgrain** [reply]
Yep that makes sense. I know this is something that's coming up fairly regularly (although not that often from my experience). I'm also not aware of a general solution or approach that has been settled on. 

I think this might be a good candidate for a projen project type (first or third party, tbd) once we have found the answers to your questions.

As a heads up: you might have more luck asking about this in other places and communities. It's certainly projen adjacent and there's a decent overlap here with CDK users, but in general I reckon this question isn't really about projen (yet).

**Comment by mrgrain** [comment]
> Would this be something to be added to projen?

If it's supporting a commonly used best practice then we can definitely add it into core. If it's something special to your use case, it's probably better placed in a custom project type.
